### ISSUE TITLE
New feature: Dropout network for bayesian approximation

### ISSUE THREAD

**Issue by isacarnekvist** [pytorch/pytorch#1508]
During eval, instead of averaging over all subtrees, you can randomly draw one dropout mask and use that for inference. This has a really nice behaviour which is that samples far away from training data naturally has more variance. See the paper: [Dropout as a Bayesian Approximation: Representing Model Uncertainty in Deep Learning](https://arxiv.org/abs/1506.02142) or read the author's [blog](http://mlg.eng.cam.ac.uk/yarin/blog.html) for really nice illustrations.

This needs the (minor) added feature to also drop during eval, and to resample the dropout mask manually. The reason to resample manually is for you to be able to draw an entire function at once, inspired by Gaussian processes.

I implemented this as a separate Dropout, called DropoutMask, but maybe it is better as a flag to Dropout? The downside of that is that there will be some special cases, like now a new Dropout object is in the backend instantiated anew for each forward pass (fix this also?), which you cannot do here since you want to keep the state.

**Comment by ethancaballero** [comment]
@smerity

**Comment by isacarnekvist** [comment]
Actually, this is much simpler to just implement as an elementwise multiplication with a bernoulli matrix in a subclass of nn.Module
